vendor/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/dynamic_certificate.go

```diff
@@ -0,0 +1,228 @@
+package server
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"reflect"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/golang/glog"
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// DynamicCertificateConfig provides a way to have dynamic https configuration for a fixed set of files.
+// You need to remember to call CheckCerts before usage and start the go routine for Run.
+type DynamicCertificateConfig struct {
+	CurrentValue atomic.Value
+
+	// LoopbackCert holds the special certificate that we create for loopback connections
+	LoopbackCert *tls.Certificate
+
+	CurrentContent DynamicCertificateContent
+
+	CertificateReferences DynamicCertificateReferences
+}
+
+type DynamicCertificateReferences struct {
+	DefaultCertificate CertKeyFileReference
+	NameToCertificate  map[string]*CertKeyFileReference
+}
+
+type CertKeyFileReference struct {
+	Cert string
+	Key  string
+}
+
+type DynamicCertificateContent struct {
+	DefaultCertificate CertKeyFileContent
+	NameToCertificate  map[string]*CertKeyFileContent
+}
+
+type CertKeyFileContent struct {
+	Cert []byte
+	Key  []byte
+}
+
+type RuntimeDynamicCertificateConfig struct {
+	Certificates      []tls.Certificate
+	NameToCertificate map[string]*tls.Certificate
+}
+
+func (c *DynamicCertificateConfig) GetCertificate(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
+	uncastObj := c.CurrentValue.Load()
+	if uncastObj == nil {
+		return nil, errors.New("tls: configuration not ready")
```

Comment:
should nit be error type that callers can retry? the error message suggests "not ready, please retry"

Reply:
> should this be an error type that callers can retry? the error message suggests "not ready, please retry"

I don't think so.  We init once before we start and should never write something naughty back in.  Also, this is a connection level error.